### PR TITLE
[cookies] Fix `--cookies-from-browser` for `snap` Firefox

### PR DESCRIPTION
**Pull request by Grub4K**
**IMPORTANT**: PRs without the template will be CLOSED

### Description of your *pull request* and other information

Implement support for `--cookies-from-browser` for `snap` Firefox and do more specific searching for the`Profiles` dirs. It also allows `os.altsep` as a path separator for `_is_path`, so that `/` is correctly categorized on Windows.

This PR has the side effect of including `glob`, as well as no longer searching infinitely deep, therefore speeding up the cookie extraction by an immense amount.

Fixes [#7520](<https://github.com/yt-dlp/yt-dlp/issues/7520>), Closes [#9011](<https://github.com/yt-dlp/yt-dlp/issues/9011>)

<details open><summary>Template</summary> 

### Before submitting a *pull request* make sure you have:
- [x] At least skimmed through [contributing guidelines](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions) including [yt-dlp coding conventions](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#yt-dlp-coding-conventions)
- [x] [Searched](https://github.com/yt-dlp/yt-dlp/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests
- [x] Checked the code with [flake8](https://pypi.python.org/pypi/flake8) and [ran relevant tests](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions)

### In order to be accepted and merged into yt-dlp each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check all of the following options that apply:
- [x] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [ ] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [ ] Fix or improvement to an extractor (Make sure to add/update tests)
- [ ] New extractor ([Piracy websites will not be accepted](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#is-the-website-primarily-used-for-piracy))
- [x] Core bug fix/improvement
- [ ] New feature (It is strongly [recommended to open an issue first](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#adding-new-feature-or-making-overarching-changes))

</details>

### CODE REVIEW

File: yt_dlp/cookies.py [hunk 1]
```diff
@@ -122,13 +123,14 @@ def _extract_firefox_cookies(profile, container, logger):
...
     if profile is None:
-        search_root = _firefox_browser_dir()
+        search_roots = list(_firefox_browser_dirs())
     elif _is_path(profile):
-        search_root = profile
+        search_roots = [profile]
     else:
-        search_root = os.path.join(_firefox_browser_dir(), profile)
+        search_roots = [os.path.join(path, profile) for path in _firefox_browser_dirs()]
+    search_root = ', '.join(map(repr, search_roots))
```
Comment (bashonly) [inline yt_dlp/cookies.py:131]: Should we raise here `if not search_roots`? the following `FileNotFoundError` won't be very helpful if `search_root` is an empty string

File: yt_dlp/cookies.py [hunk 2]
```diff
@@ -182,12 +184,21 @@ def _extract_firefox_cookies(profile, container, logger):
...
-    return os.path.expanduser('~/.mozilla/firefox')
+        yield os.path.expanduser('~/Library/Application Support/Firefox/Profiles')
+
+    else:
+        yield from map(os.path.expanduser, ('~/.mozilla/firefox', '~/snap/firefox/common/.mozilla/firefox'))
+
+
+def _firefox_cookie_dbs(roots):
+    for root in map(os.path.abspath, roots):
+        for pattern in '', '*/', 'Profiles/*/':
```
Comment (bashonly) [inline yt_dlp/cookies.py:200]: helps w/ readability for me
```suggestion
        for pattern in ('', '*/', 'Profiles/*/'):
```

**Review by bashonly — CHANGES_REQUESTED** [review]

File: yt_dlp/cookies.py (hunk 1 above)
Comment (Grub4K) [reply yt_dlp/cookies.py:131]: `search_roots` will always contain at least one entry (`_firefox_browser_dirs` yields at least one item and `[profile]`)

**Review by bashonly — APPROVED** [review]

**Comment by mmafic** [comment]
Is it possible to add support for custom profile directories? Firefox allows you to choose a location when creating a profile, the list of all profiles is stored in profile.ini, which is in AppData/Roaming on Windows.

**Comment by bashonly** [comment]
> Is it possible to add support for custom profile directories?

you can already pass any profile directory path, e.g.
`--cookies-from-browser "firefox:/path/to/profile"`

**Comment by Grub4K** [comment]
> ...the list of all profiles is stored in profile.ini, which is in AppData/Roaming on Windows.

This is not a bad idea. I will create a POC that does this to see if it makes sense to implement.

**Pull request merged**